core/src/main/scala/org/typelevel/ci/CIString.scala

```diff
@@ -24,52 +24,46 @@ import scala.math.Ordered
 
 /** A case-insensitive String.
   *
-  * Two CI strings are equal if and only if they are the same length, and each corresponding
-  * character is equal after calling either `toUpper` or `toLower`.
+  * Comparisions are based on the case folded representation of the `String`
+  * as defined by the Unicode standard. See [[CaseFoldedString]] for a full
+  * discussion on those rules.
   *
-  * Ordering is based on a string comparison after folding each character to uppercase and then back
-  * to lowercase.
-  *
-  * All comparisons are insensitive to locales.
+  * @note This class differs from [[CaseFoldedString]] in that it keeps a
+  *       reference to original input `String` in whatever form it was
+  *       given. This makes [[CIString]] useful if you which to perform case
+  *       insensitive operations on a `String`, but then recover the original,
+  *       unaltered form. If you do not care about the original input form,
+  *       and just want a single case insensitive `String` value, then
+  *       [[CaseFoldedString]] is more efficient and you should consider using
+  *       that directly.
   *
   * @param toString
   *   The original value the CI String was constructed with.
   */
-final class CIString private (override val toString: String)
+final class CIString private (override val toString: String, val asCaseFoldedString: CaseFoldedString)
```

Comment:
@rossabaker yeah, that makes sense. I actually benchmarked def/lazy val/val. lazy val seemed like a reasonable tradeoff. I was hesitant because I think a lot of our use cases are hashCode/equality heavy, but I can make it a `lazy val` or use the null-init pattern.

Reply:
If you can stomach the var, null-init is the fastest unless multiple threads initialize it concurrently, which should be rare.  This is how`String#hashCode` operates.